### ISSUE TITLE
Improve `@resource` to allow (real) asynchronous retrieval only if needed

### ISSUE THREAD

**Issue by HaloWang** [Tampermonkey/tampermonkey#1461]
I find if I load a large(20mb+) file by tag `@resource`. It will (block/make it slow) other running code on my webpage especially scripts added by tampermonkey.

Could I load my lagre file async asynchronously?

I try to use a local http-server. But some website like Github refuse my HttpRequest: 

> Refused to connect to 'http://localhost:8000' because it violates the following Content Security Policy directi.....

My codes in tampermonkey in Chrome:

```js
// ==UserScript==
// @name         XXXX
// @version      XXXX
// @description  XXXX
// @icon         XXXX
// @match        XXXX
// @match        XXXX
// @require               file:///XXXX\index.js
// @resource     BIGJson  file:///XXXX\dict.json
// @grant        GM_addStyle
// @grant        GM_getResourceText
// ==/UserScript==

// Your code here...
// But there is addtional code. I write my logic in tag @require.

```

**Comment by 7nik** [comment]
There is async GM API: `GM.getResourceText` but IDK whether it is more efficient. Anyway passing and parsing 20mb+ text is a hard thing.

> Refused to connect

Use [GM_xmlhttpRequest](https://www.tampermonkey.net/documentation.php?version=4.14&ext=dhdg#GM_xmlhttpRequest) instead of fetch/XHR.

**Comment by HaloWang** [comment]
Thank you @7nik 

I find this codes work for me. `GM_xmlhttpRequest` can help me, even I can not track my request in Chrome Devtools.

```js
// ==UserScript==
// @name         Test what @7nik said
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  try to figure out problems
// @author       You
// @match        https://github.com/Tampermonkey/tampermonkey/issues/1461
// @icon         https://www.google.com/s2/favicons?sz=64&domain=github.com
// @grant        GM_xmlhttpRequest
// @connect      localhost
// ==/UserScript==

;(function () {
  'use strict'
  GM_xmlhttpRequest({
    url: 'http://localhost:8000',
    onload: response => {
      if (response.status === 200) {
        console.log('✅ success!')
        console.log(response.responseText)
      } else {
      }
    },
    onerror: err => {
      console.log('💥', err)
    },
  })
})()

```

By the way, i mean adding a config tag like following code will block js invocation:

```js
// @resource  BIGJson  file:///XXXX\dict_size_30MB.json
```

Even I haven't use `GM_getResourceText` to get it. 🤪 I will drop this method.

**Comment by derjanb** [comment]
Reopneing since there is room for improvement

**Comment by derjanb** [comment]
@HaloWang the page blocking issue was fixed in #1486

**Comment by HaloWang** [comment]
> @HaloWang the page blocking issue was fixed in #1486

I'm now using httpRequest but thanks for your excellent works.

**Comment by raszpl** [comment]
> `GM_xmlhttpRequest` helped me, even I can not track my request in Chrome Devtools.

because fetch happens in tampermonkey execution context, if you really want to see it in devtools you need to go "manage extensions" and then in tampermonkey "Inspect views background.html"
